### PR TITLE
Adds test scenarios for DNSSEC10

### PR DESCRIPTION
**Pull request by matsduf**
## Purpose

The specification of DNSSEC10 has been rewritten in PR #1179 (not yet merged or implemented). This PR defines test scenarios for DNSSEC10 based on the mentioned PR. This PR is not yet complete since all test zones are not yet created.

## Context

#1179

## Status of PR

The content of the PR is complete. There are two scenarios for which it does not seem to be possible to create test zones with the current tools. My plan is to rearrange the commits and also review what to save in Git. The keys for the Bind zone can probably be left out and recreated each time. Some documentation around Bind and its files will also be updated. None of this affects the test zones for the unit tests.

## Changes

All planned scenarios have been defined. Ticked scenarios have complete test zone data and can be used for preliminary unit tests.

* [x] GOOD-NSEC-1
* [x] GOOD-NSEC3-1
* [x] ALGO-NOT-SUPP-BY-ZM-1                                                                                                                                                                                                 
* [x] ALGO-NOT-SUPP-BY-ZM-2
* [x] ERR-MULT-NSEC-1                                                                                                                                                                                                       
* [x] ERR-MULT-NSEC3-1
* [x] EXP-NSEC-NSEC3-MISS-1
* [x] INCONSISTENT-NSEC-1
* [x] INCONSISTENT-NSEC3-1
* [x] INCONSIST-NSEC-NSEC3-1                                                                                                                                                                                                
* [x] INCONSIST-NSEC-NSEC3-2
* [x] MIXED-NSEC-NSEC3-1                                                                                                                                                                                                    
* [x] MIXED-NSEC-NSEC3-1
* [x] NSEC3PARAM-GIVES-ERR-ANSWER-1
* [x] NSEC3PARAM-GIVES-ERR-ANSWER-2
* [x] NSEC3PARAM-GIVES-ERR-ANSWER-3
* [x] NSEC3PARAM-Q-RESPONSE-ERR-1
* [x] NSEC3PARAM-Q-RESPONSE-ERR-2
* [ ] NSEC3PARAM-Q-RESPONSE-ERR-3 (maybe not possible in coredns()?)
* [x] NSEC3-ERR-TYPE-LIST-1
* [x] NSEC3-ERR-TYPE-LIST-2
* [x] NSEC3-MISMATCHES-APEX-1                                                                                                                                                                                               
* [x] NSEC3-MISSING-SIGNATURE-1
* [x] NSEC3-NODATA-MISSING-SOA-1
* [x] NSEC3-NODATA-WRONG-SOA-1
* [x] NSEC3-NO-VERIFIED-SIGNATURE-1
* [x] NSEC3-NO-VERIFIED-SIGNATURE-2                                                                                                                                                                                         
* [x] NSEC3-NO-VERIFIED-SIGNATURE-3
* [x] NSEC3-NO-VERIFIED-SIGNATURE-4
* [x] NSEC-ERR-TYPE-LIST-1
* [x] NSEC-ERR-TYPE-LIST-2
* [x] NSEC-GIVES-ERR-ANSWER-1
* [x] NSEC-GIVES-ERR-ANSWER-2
* [x] NSEC-MISMATCHES-APEX-1
* [x] NSEC-MISMATCHES-APEX-2                                                                                                                                                                                                
* [x] NSEC-MISSING-SIGNATURE-1
* [x] NSEC-NODATA-MISSING-SOA-1
* [x] NSEC-NODATA-WRONG-SOA-1
* [x] NSEC-NO-VERIFIED-SIGNATURE-1
* [x] NSEC-NO-VERIFIED-SIGNATURE-2                                                                                                                                                                                          
* [x] NSEC-NO-VERIFIED-SIGNATURE-3
* [x] NSEC-NO-VERIFIED-SIGNATURE-4
* [x] NSEC-QUERY-RESPONSE-ERR-1
* [x] NSEC-QUERY-RESPONSE-ERR-2
* [ ] NSEC-QUERY-RESPONSE-ERR-3 (maybe not possible in coredns()?)
* [x] SERVER-NO-DNSSEC-1                                                                                                                                                                                                    
* [x] SERVER-NO-DNSSEC-2
* [x] ZONE-NO-DNSSEC-1

## How to test this PR

Review the test scenarios. Files under the `Bind` directory are not meaningful to review.

### CODE REVIEW

**Comment by matsduf** [comment]
@tgreenx, all test zones are created for updated DNSSEC10.

File: docs/public/specifications/test-zones/DNSSEC-TP/dnssec10.md [hunk 1]
```diff
@@ -0,0 +1,560 @@
...
+
+* Zone: nsec3-err-type-list-2.dnssec10.xa
+  * The type list of the NSEC3 record misses RRSIG.
+
+### NSEC3-MISMATCHES-APEX-1
+An NSEC3 zone. The hash owner name of the NSEC3 record is errouneous.
+
+* Zone: nsec3-mismatches-apex-1.dnssec10.xa
+  * The hash owner name of the NSEC3 record in response to the NSEC query is
+    errouneous and does not match apex.
```
Comment (marc-vanderwal) [inline docs/public/specifications/test-zones/DNSSEC-TP/dnssec10.md:349]:
```suggestion
An NSEC3 zone. The hash owner name of the NSEC3 record is erroneous.

* Zone: nsec3-mismatches-apex-1.dnssec10.xa
  * The hash owner name of the NSEC3 record in response to the NSEC query is
    erroneous and does not match apex.
```

**Review by marc-vanderwal — COMMENTED** [review]
I see no obvious issues except for a little typo.

**Comment by matsduf** [comment]
@tgreenx and @marc-vanderwal, please review.

File: test-zone-data/DNSSEC-TP/dnssec10/Bind/README.md [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+* [Key generation and zone signing](#key-generation-and-zone-signing)
+* [Creating a variant zone](#creating-a-variant-zone)
+* [Synchronizing the keys and zones at checkout](#synchronizing-the-keys-and-zones-at-checkout)
+* [Starting and stopping Bind](#starting-and-stopping-bind)
+* [Reloading Bind](#reloading-bind)
+* [Checking Bind](#checking-bind)
+
+## Variable below
+
+Below `%ZONE-NAME%` means the name of the zone in question, without the traling
```
Comment (marc-vanderwal) [inline test-zone-data/DNSSEC-TP/dnssec10/Bind/README.md:16]:
```suggestion
Below `%ZONE-NAME%` means the name of the zone in question, without the trailing
```

**Review by marc-vanderwal — COMMENTED** [review]
Looks fine. I spotted just one more typo.

**Comment by matsduf** [comment]
@marc-vanderwal and @tgreenx, please review.

**Review by marc-vanderwal — APPROVED** [review]

File: docs/public/specifications/test-zones/DNSSEC-TP/dnssec10.md [hunk 3]
```diff
@@ -0,0 +1,570 @@
...
+:------------------------------|:-----------------------------------------------------------------------------|:--------------------
+GOOD-NSEC-1                    | DS10_HAS_NSEC                                                                | 2)
+GOOD-NSEC3-1                   | DS10_HAS_NSEC3                                                               | 2)
+ALGO-NOT-SUPP-BY-ZM-1          | DS10_ALGO_NOT_SUPPORTED_BY_ZM, DS10_HAS_NSEC                                 | 2)
+ALGO-NOT-SUPP-BY-ZM-2          | DS10_ALGO_NOT_SUPPORTED_BY_ZM, DS10_HAS_NSEC3                                | 2)
+ERR-MULT-NSEC-1                | DS10_ERR_MULT_NSEC, DS10_HAS_NSEC                                            | 2)
+ERR-MULT-NSEC-2                | DS10_ERR_MULT_NSEC, DS10_HAS_NSEC                                            | 2)
+ERR-MULT-NSEC3-1               | DS10_ERR_MULT_NSEC3, DS10_HAS_NSEC3                                          | 2)
+EXP-NSEC-NSEC3-MISS-1          | DS10_EXPECTED_NSEC_NSEC3_MISSING                                             | 2)
+INCONSISTENT-NSEC-1            | DS10_INCONSISTENT_NSEC, DS10_HAS_NSEC                                        | 2)
```
[the remark this answers is not in the document]
Comment (matsduf) [reply docs/public/specifications/test-zones/DNSSEC-TP/dnssec10.md:98]: That is a mistake in the specification. It should continue with next "section" ("Send _NSEC3PARAM Query_ to the name server IP") or else the order between the NSEC query and the NSEC3PARAM query matters, which is confusing and bad. I'll fix the specification.

[the remark this answers is not in the document]
Comment (matsduf) [reply docs/public/specifications/test-zones/DNSSEC-TP/dnssec10.md:98]: See https://github.com/zonemaster/zonemaster/pull/1179#issuecomment-2478502003 and https://github.com/zonemaster/zonemaster/commit/73836747a81d213297ad88d4d849c988f1817403

File: docs/public/specifications/test-zones/DNSSEC-TP/dnssec10.md [hunk 4]
```diff
@@ -0,0 +1,570 @@
...
+INCONSISTENT-NSEC3-1           | DS10_INCONSISTENT_NSEC3, DS10_HAS_NSEC3                                      | 2)
```
[the remark this answers is not in the document]
Comment (matsduf) [reply docs/public/specifications/test-zones/DNSSEC-TP/dnssec10.md:99]: It should be corrected by corrected specification.

**Review by marc-vanderwal — APPROVED** [review]

File: docs/public/specifications/test-zones/DNSSEC-TP/dnssec10.md [hunk 5]
```diff
@@ -0,0 +1,570 @@
...
+INCONSIST-NSEC-NSEC3-1         | DS10_INCONSISTENT_NSEC_NSEC3                                                 | 2)
+INCONSIST-NSEC-NSEC3-2         | DS10_INCONSISTENT_NSEC_NSEC3, DS10_INCONSISTENT_NSEC, DS10_INCONSISTENT_NSEC3| 2)
+MIXED-NSEC-NSEC3-1             | DS10_MIXED_NSEC_NSEC3                                                        | 2)
+MIXED-NSEC-NSEC3-1             | DS10_MIXED_NSEC_NSEC3                                                        | 2)
+NSEC3PARAM-GIVES-ERR-ANSWER-1  | DS10_NSEC3PARAM_GIVES_ERR_ANSWER, DS10_HAS_NSEC3                             | 2)
+NSEC3PARAM-GIVES-ERR-ANSWER-2  | DS10_NSEC3PARAM_GIVES_ERR_ANSWER, DS10_EXPECTED_NSEC_NSEC3_MISSING           | 2)
+NSEC3PARAM-GIVES-ERR-ANSWER-3  | DS10_NSEC3PARAM_GIVES_ERR_ANSWER, DS10_HAS_NSEC3                             | 2)
+NSEC3PARAM-Q-RESPONSE-ERR-1    | DS10_NSEC3PARAM_QUERY_RESPONSE_ERR, DS10_HAS_NSEC3                           | 2)
+NSEC3PARAM-Q-RESPONSE-ERR-2    | DS10_NSEC3PARAM_QUERY_RESPONSE_ERR, DS10_HAS_NSEC3                           | 2)
```
[the remark this answers is not in the document]
Comment (matsduf) [reply docs/public/specifications/test-zones/DNSSEC-TP/dnssec10.md]: Fixed.

**Comment by matsduf** [comment]
Commit 1aaab825b81323627f7ef43d585a99352babb075 adds scenarios ERR-MULT-NSEC3PARAM-1 and MISMATCHES-APEX-1. Scenario GIVES-ERR-ANSWER-3 has been removed.

File: docs/public/specifications/test-zones/DNSSEC-TP/dnssec10.md [hunk 6]
```diff
@@ -0,0 +1,582 @@
...
+
+### NSEC-QUERY-RESPONSE-ERR-2
+An NSEC zone. Error in response to NSEC query.
+
+* Zone: nsec-query-response-err-2.dnssec10.xa
+  * The response on the NSEC query has the [RCODE Name] "REFUSED".
+
+### NSEC-QUERY-RESPONSE-ERR-3
+An NSEC zone. Error in response to NSEC query on ns1. No NSEC or NSEC3 in
+responsess from ns2.
```
[the remark this answers is not in the document]
Comment (matsduf) [reply docs/public/specifications/test-zones/DNSSEC-TP/dnssec10.md:542]: Fixed.

**Comment by matsduf** [comment]
> The file docs/public/SUMMARY.md should be updated too.

Fixed. That is easy to miss. We need a script to search for unlinked documents.

**Pull request merged**